### PR TITLE
Updated kafka versions to 1.0.0

### PR DESCRIPTION
**Pull request by mithuns**
Updated kafka versions to 1.0.0
This PR solves this issue
https://github.com/simplesteph/kafka-streams-course/issues/1
Also, had to update gitignore to exclude the .project and .classpath files being created inside as well as at the base repo level. 

Let me know if something more is needed.

### CODE REVIEW

**Comment by simplesteph** [comment]
Hi @mithuns 

Thanks for the PR. This actually updates dependencies, but brings in a lot of the deprecation warning due to API changes.

Do you want to give it a crack at fixing those? It'll be good for learning. 

Thanks!
Stephane

**Review by simplesteph — CHANGES_REQUESTED** [review]
will also need to update the scala bit - if you don't know scala no worries I'll do it myself

**Comment by mithuns** [comment]
Can we do the scala bit in a separate PR ?

**Comment by simplesteph** [comment]
sure, but please make sure you resolve conflicts in this one and don't edit the travis file

**Comment by mithuns** [comment]
I am facing a few issues, also I have never tried to build a scala project in eclipse before. 
So, you can take this over and update that part accordingly or I can take a shot at this over the weekend.

**Comment by simplesteph** [comment]
Don't worry about the scala part I'll take care of it :) try just fixing the depreciation in java

**Comment by simplesteph** [comment]
@mithuns  please target 1.0.0 for your PR. I want to maintain the 1.0.0 efforts in a separate branch for now :)

**Comment by simplesteph** [comment]
hi @mithuns , any plans on finishing this PR? Otherwise I'll finish it this week

**Comment by mithuns** [comment]
Hey @simplesteph, I have updated code, please review.

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 1]
```diff
@@ -33,18 +32,13 @@ public static void main(String[] args) {
...
-        final Deserializer<JsonNode> jsonDeserializer = new JsonDeserializer();
-        final Serde<JsonNode> jsonSerde = Serdes.serdeFrom(jsonSerializer, jsonDeserializer);
-
-
-        KStreamBuilder builder = new KStreamBuilder();
-
-        KStream<String, JsonNode> bankTransactions =
-                builder.stream(Serdes.String(), jsonSerde, "bank-transactions");
-
+        config.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.serdeFrom(JsonSerializer.class));
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java]: why did you remove the SerDes and add default? I think it's fine, but we want the videos to keep in line with what we already have, so the previous code should be fine

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 2]
```diff
@@ -53,22 +47,21 @@ public static void main(String[] args) {
         initialBalance.put("time", Instant.ofEpochMilli(0L).toString());
 
         KTable<String, JsonNode> bankBalance = bankTransactions
-                .groupByKey(Serdes.String(), jsonSerde)
+                .groupByKey(Serialized.with(Serdes.serdeFrom(String.class),Serdes.serdeFrom(JsonNode.class)))
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java]: Why can't you use `Serdes.String()`? 
Again, externalise `jsonSerde` like we had before and keep the code. only `Serialized.with` is added

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 3]
```diff
@@ -53,22 +47,21 @@ public static void main(String[] args) {
...
                 .aggregate(
                         () -> initialBalance,
                         (key, transaction, balance) -> newBalance(transaction, balance),
-                        jsonSerde,
-                        "bank-balance-agg"
+                        Materialized.as("bank-balance-agg")
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java:64]: keep the `jsonSerde`, this line is good though

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 4]
```diff
@@ -53,22 +47,21 @@ public static void main(String[] args) {
...
                 );
 
-        bankBalance.to(Serdes.String(), jsonSerde,"bank-balance-exactly-once");
+        bankBalance.toStream();
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java]: `toStream().to(...)`. Right now you removed the writing to the output topic

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 5]
```diff
@@ -53,22 +47,21 @@ public static void main(String[] args) {
...
 
-        KafkaStreams streams = new KafkaStreams(builder, config);
+        KafkaStreams streams = new KafkaStreams(builder.build(), config);
         streams.cleanUp();
         streams.start();
 
         // print the topology
-        System.out.println(streams.toString());
+        System.out.println(streams.localThreadsMetadata());
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java]: `streams.localThreadMetadata.foreach(System.out.println)` or something like that

File: favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java [hunk 6]
```diff
@@ -50,18 +50,18 @@ public static void main(String[] args) {
...
 
-        KafkaStreams streams = new KafkaStreams(builder, config);
+        KafkaStreams streams = new KafkaStreams(builder.build(), config);
         // only do this in dev - not in prod
         streams.cleanUp();
         streams.start();
 
         // print the topology
-        System.out.println(streams.toString());
+        System.out.println(streams.localThreadsMetadata());
```
Comment (simplesteph) [inline favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java]: same comment here

File: streams-starter-project/src/main/java/com/github/simplesteph/udemy/kafka/streams/StreamsStarterApp.java [hunk 7]
```diff
@@ -20,18 +20,18 @@ public static void main(String[] args) {
...
         kStream.to("word-count-output");
 
-        KafkaStreams streams = new KafkaStreams(builder, config);
+        KafkaStreams streams = new KafkaStreams(builder.build(), config);
         streams.cleanUp(); // only do this in dev - not in prod
         streams.start();
 
         // print the topology
-        System.out.println(streams.toString());
+        System.out.println(streams.localThreadsMetadata());
```
Comment (simplesteph) [inline streams-starter-project/src/main/java/com/github/simplesteph/udemy/kafka/streams/StreamsStarterApp.java]: same here

File: user-event-enricher/src/main/java/com/github/simplesteph/udemy/kafka/streams/UserEventEnricherApp.java [hunk 8]
```diff
@@ -55,12 +55,12 @@ public static void main(String[] args) {
...
 
 
-        KafkaStreams streams = new KafkaStreams(builder, config);
+        KafkaStreams streams = new KafkaStreams(builder.build(), config);
         streams.cleanUp(); // only do this in dev - not in prod
         streams.start();
 
         // print the topology
-        System.out.println(streams.toString());
+        System.out.println(streams.localThreadsMetadata());
```
Comment (simplesteph) [inline user-event-enricher/src/main/java/com/github/simplesteph/udemy/kafka/streams/UserEventEnricherApp.java]: same here

File: word-count/src/main/java/com/github/simplesteph/udemy/kafka/streams/WordCountApp.java [hunk 9]
```diff
@@ -36,18 +38,18 @@ public static void main(String[] args) {
...
 
-        KafkaStreams streams = new KafkaStreams(builder, config);
+        KafkaStreams streams = new KafkaStreams(builder.build(), config);
         streams.start();
 
         // Update:
         // print the topology every 10 seconds for learning purposes
         while(true){
-            System.out.println(streams.toString());
+            System.out.println(streams.localThreadsMetadata());
```
Comment (simplesteph) [inline word-count/src/main/java/com/github/simplesteph/udemy/kafka/streams/WordCountApp.java]: same here

**Review by simplesteph — CHANGES_REQUESTED** [review]
see comments

**Comment by simplesteph** [comment]
thanks @mithuns , see comments :) I'll review the code again afterwards

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java (hunk 1 above)
Comment (mithuns) [reply bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java]: I believe I followed the example given here
https://kafka.apache.org/10/documentation/streams/

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java (hunk 3 above)
Comment (mithuns) [reply bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java:64]: All the versions of aggregate that will take jsonSerde, are deprecated.
[KGroupedStream](https://kafka.apache.org/10/javadoc/org/apache/kafka/streams/kstream/KGroupedStream.html)

File: favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java [hunk 10]
```diff
@@ -50,18 +50,18 @@ public static void main(String[] args) {
         KTable<String, Long> favouriteColours = usersAndColoursTable
                 // 5 - we group by colour within the KTable
                 .groupBy((user, colour) -> new KeyValue<>(colour, colour))
-                .count("CountsByColours");
+                .count(Materialized.as("CountsByColours"));
```
Comment (simplesteph) [inline favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java:53]: This one does not work, I ran the example and it fails at runtime with a serialisation error. Remember the default serialiser is String and now you get a Long.

This one was a bit painful to figure out, but use this:

```
Serde<String> stringSerde = Serdes.String();
        Serde<Long> longSerde = Serdes.Long();

        // step 3 - we count the occurences of colours
        KTable<String, Long> favouriteColours = usersAndColoursTable
                // 5 - we group by colour within the KTable
                .groupBy((user, colour) -> new KeyValue<>(colour, colour))
                .count(Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as("CountsByColours")
                        .withKeySerde(stringSerde)
                        .withValueSerde(longSerde));
```

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 11]
```diff
@@ -33,18 +38,17 @@ public static void main(String[] args) {
...
-
-        KStreamBuilder builder = new KStreamBuilder();
-
-        KStream<String, JsonNode> bankTransactions =
-                builder.stream(Serdes.String(), jsonSerde, "bank-transactions");
-
+        
+        StreamsBuilder builder = new StreamsBuilder();
+        
+        KStream<String, JsonNode> bankTransactions = builder.stream("bank-transactions");
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java:49]: you dropped the serializers here, this will fail at runtime

use this instead:
```
KStream<String, JsonNode> bankTransactions = builder.stream("bank-transactions", 
                Consumed.with(Serdes.String(), jsonSerde));
```

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java (hunk 3 above)
Comment (simplesteph) [reply bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java:64]: As same comment as below something like this should work:
```
Materialized.<String, JsonNode, KeyValueStore<Bytes, byte[]>>as("bank-balance-agg")
                                    .withKeySerde(Serdes.String())
                                    .withValueSerde(jsonSerde)
```

File: bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java [hunk 12]
```diff
@@ -53,22 +57,21 @@ public static void main(String[] args) {
...
                 .aggregate(
                         () -> initialBalance,
                         (key, transaction, balance) -> newBalance(transaction, balance),
-                        jsonSerde,
-                        "bank-balance-agg"
+                        Materialized.as("bank-balance-agg")
                 );
 
-        bankBalance.to(Serdes.String(), jsonSerde,"bank-balance-exactly-once");
+        bankBalance.toStream().to("bank-balance-exactly-once");
```
Comment (simplesteph) [inline bank-balance-exactly-once/src/main/java/com/github/simplesteph/udemy/kafka/streams/BankBalanceExactlyOnceApp.java:67]:
```
        bankBalance.toStream().to("bank-balance-exactly-once", Produced.with(Serdes.String(), jsonSerde));
```

**Review by simplesteph — CHANGES_REQUESTED** [review]
So I think you got the idea, but a few more comments. Some API may seem un-intuitive (due to Java...)

I've added the solutions for many of the problems. Please submit a correct PR.

Also, please run each example with sample data as in the course and make sure the application is actually running fine. That's how I detected a few issues :)

File: favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java (hunk 10 above)
Comment (mithuns) [reply favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java:53]: I will update accordingly, I dont have access to course's sample test-data though.

Comment (simplesteph) [reply favourite-colour-java/src/main/java/com/github/simplesteph/udemy/kafka/streams/FavouriteColourApp.java:53]: No need for sample data, just push one sample message and if there's no runtime error it's good ! You can easily figure out what data might look like :)

**Comment by simplesteph** [comment]
closed in favour of #10
